### PR TITLE
[Bulk mode] Limit the size of bulker

### PR DESCRIPTION
**Pull request by shi-su**
**What I did**
Limit the size of the bulker.

**Why I did it**
Without a constraint on the size of bulkers, the bulked operation may take longer than the SAI timeout to finish. Such behavior would result in errors and swss crash. This PR limits the bulk size and avoids SAI timeout in bulk mode.

**How I verified it**

**Details if related**

### CODE REVIEW

File: orchagent/bulker.h [hunk 1]
```diff
@@ -280,29 +282,19 @@ class EntityBulker
...
-                SWSS_LOG_ERROR("EntityBulker.flush remove entries failed, number of entries to remove: %zu, status: %s",
-                                removing_entries.size(), sai_serialize_status(status).c_str());
-            }
 
-            for (size_t ir = 0; ir < count; ir++)
-            {
-                auto& entry = rs[ir];
-                sai_status_t *object_status = removing_entries[entry];
-                if (object_status)
+                if (rs.size() >= MAX_BULK_SIZE)
```
Comment (qiluo-msft) [inline orchagent/bulker.h:286]: >if [](http://example.com/codeflow?start=16&length=2)

You can move this if-block into inside of above if-block

File: orchagent/bulker.h [hunk 2]
```diff
@@ -455,6 +429,93 @@ class EntityBulker
...
+            sai_status_t *object_status = removing_entries[entry];
+            if (object_status)
+            {
+                *object_status = statuses[ir];
+            }
+        }
+    }
+
+    void flush_creating_entries(
+        _In_ std::vector<Te> rs,
```
Comment (qiluo-msft) [inline orchagent/bulker.h:461]: >vector [](http://example.com/codeflow?start=18&length=6)

Copy a vector is too heavy. Could you use reference?

File: orchagent/bulker.h [hunk 3]
```diff
@@ -372,31 +355,22 @@ class EntityBulker
                         ts.push_back(attr);
                         status_vector.push_back(object_status);
                     }
+
+                    if (rs.size() >= MAX_BULK_SIZE)
+                    {
+                        flush_setting_entries(rs, ts, status_vector);
+                        rs.clear();
+                        ts.clear();
+                        status_vector.clear();
```
Comment (qiluo-msft) [inline orchagent/bulker.h:364]: If you use reference in above function, you can move the 3 clear() into the above function.

File: orchagent/bulker.h [hunk 4]
```diff
@@ -455,6 +429,93 @@ class EntityBulker
...
     typename Ts::bulk_remove_entry_fn                       remove_entries;
     typename Ts::bulk_set_entry_attribute_fn                set_entries_attribute;
+
+    void flush_removing_entries(
+        _In_ std::vector<Te> rs)
+    {
+        size_t count = rs.size();
+        std::vector<sai_status_t> statuses(count);
+        sai_status_t status = (*remove_entries)((uint32_t)count, rs.data(), SAI_BULK_OP_ERROR_MODE_IGNORE_ERROR, statuses.data());
+        if (status == SAI_STATUS_SUCCESS)
```
Comment (qiluo-msft) [inline orchagent/bulker.h:439]: >status [](http://example.com/codeflow?start=12&length=6)

The status if useful, let's return it.

File: orchagent/bulker.h [hunk 5]
```diff
@@ -280,29 +282,19 @@ class EntityBulker
...
-            for (size_t ir = 0; ir < count; ir++)
-            {
-                auto& entry = rs[ir];
-                sai_status_t *object_status = removing_entries[entry];
-                if (object_status)
+                if (rs.size() >= MAX_BULK_SIZE)
                 {
-                    *object_status = statuses[ir];
+                    flush_removing_entries(rs);
+                    rs.clear();
```
Comment (qiluo-msft) [inline orchagent/bulker.h:289]: >rs.clear(); [](http://example.com/codeflow?start=20&length=11)

If you use reference in above function, you can move the clear() into the above function.

File: orchagent/bulker.h [hunk 6]
```diff
@@ -280,29 +282,19 @@ class EntityBulker
...
                 }
             }
+
+            if (!rs.empty())
```
Comment (qiluo-msft) [inline orchagent/bulker.h:293]: >if (!rs.empty()) [](http://example.com/codeflow?start=12&length=16)

you can move the check into flush function.

File: orchagent/bulker.h [hunk 7]
```diff
@@ -11,6 +11,8 @@
 #include "logger.h"
 #include "sai_serialize.h"
 
+#define MAX_BULK_SIZE 1000
```
Comment (lguohan) [inline orchagent/bulker.h:14]: can we make this configurable? maybe via cmd line option or configuration file?

File: orchagent/bulker.h (hunk 1 above)
Comment (shi-su) [reply orchagent/bulker.h]: Moved it to the if-block above.

File: orchagent/bulker.h (hunk 2 above)
Comment (shi-su) [reply orchagent/bulker.h]: Passing reference instead.

File: orchagent/bulker.h (hunk 3 above)
Comment (shi-su) [reply orchagent/bulker.h]: Moved the clear() into the function.

File: orchagent/bulker.h (hunk 5 above)
Comment (shi-su) [reply orchagent/bulker.h]: Moved the clear() into the function.

File: orchagent/bulker.h (hunk 4 above)
Comment (shi-su) [reply orchagent/bulker.h:424]: Updated the functions to return the status.

File: orchagent/bulker.h (hunk 6 above)
Comment (shi-su) [reply orchagent/bulker.h]: Moved the check for empty into the flush function.

File: orchagent/bulker.h (hunk 7 above)
Comment (shi-su) [reply orchagent/bulker.h]: Added a command-line option to configure the max bulk size.

File: orchagent/main.cpp [hunk 8]
```diff
@@ -397,6 +401,9 @@ int main(int argc, char **argv)
                 sairedis_rec_filename = optarg;
             }
             break;
+        case 'k':
+            gMaxBulkSize = atoi(optarg);
```
Comment (qiluo-msft) [inline orchagent/main.cpp:405]: >gMaxBulkSize [](http://example.com/codeflow?start=12&length=12)

Please handle error and corner cases.

File: orchagent/bulker.h [hunk 9]
```diff
@@ -11,6 +11,8 @@
 #include "logger.h"
 #include "sai_serialize.h"
 
+extern size_t gMaxBulkSize;
```
Comment (qiluo-msft) [inline orchagent/bulker.h:14]: >gMaxBulkSize [](http://example.com/codeflow?start=14&length=12)

Better to define in below classes, also its default value.

File: tests/mock_tests/mock_orchagent_main.h [hunk 10]
```diff
@@ -25,6 +25,7 @@ extern bool gLogRotate;
 extern bool gSaiRedisLogRotate;
 extern ofstream gRecordOfs;
 extern string gRecordFile;
+extern size_t gMaxBulkSize;
```
Comment (qiluo-msft) [inline tests/mock_tests/mock_orchagent_main.h:28]: >gMaxBulkSize [](http://example.com/codeflow?start=14&length=12)

Try remove global var from test.

Comment (shi-su) [reply tests/mock_tests/mock_orchagent_main.h]: Removed from tests.

File: orchagent/main.cpp (hunk 8 above)
Comment (shi-su) [reply orchagent/main.cpp]: Added a logic to handle errors and corner cases.

File: orchagent/bulker.h (hunk 9 above)
Comment (shi-su) [reply orchagent/bulker.h]: Defined a member variable in the classes for the max bulk size and setting the value in the orch that creates the bulker. As the default is currently global, the default value is not added to the class.

**Review by qiluo-msft — APPROVED** [review]

**Comment by yxieca** [comment]
@shi-su  please help open PR for 202012 branch. Thanks.
